### PR TITLE
ARTEMIS-1273 Bounded OrderedExecutor

### PR DESCRIPTION
**Pull request by franz1981**
Introduce by default a Bounded OrderedExecutor that:
1. uses Array based lock/wait-free queues to have contiguos allocations
2. OOM protected limiting the allowed max size of the queue
3. produce garbage only if the consumer can't keep up with the producer by a configurable amount
4. uses a scalable handmade spin lock (XADD based)
5. cost of monitoring pending tasks list size O(1)
6. allow optimized versions of task submission using different specialized concurrent queues

### CODE REVIEW

File: artemis-commons/src/main/java/org/apache/activemq/artemis/utils/BoundedOrderedExecutorFactory.java [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+import org.jctools.queues.MpscArrayQueue;
+import org.jctools.queues.MpscChunkedArrayQueue;
+import org.jctools.queues.SpscArrayQueue;
+import org.jctools.queues.SpscChunkedArrayQueue;
+import org.jctools.util.Pow2;
+
+/**
+ * A factory for producing executors that run all tasks in order, which delegate to a single common executor instance.
+ */
+public final class BoundedOrderedExecutorFactory implements ExecutorFactory {
```
Comment (clebertsuconic) [inline artemis-commons/src/main/java/org/apache/activemq/artemis/utils/BoundedOrderedExecutorFactory.java]: where the word Bounded came from? is this a good name?

What about ArrayOrderedExecutor?

Comment (clebertsuconic) [inline artemis-commons/src/main/java/org/apache/activemq/artemis/utils/BoundedOrderedExecutorFactory.java]: QueueArrayOrderedExecutor

File: artemis-commons/src/main/java/org/apache/activemq/artemis/utils/BoundedOrderedExecutorFactory.java [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+         } else {
+            return false;
+         }
+      }
+
+      @Override
+      public void execute(Runnable command) {
+         //no optimisations on recursive offers
+         if (!commands.offer(command)) {
+            throw new RejectedExecutionException("can't submit the task: max capacity reached");
```
Comment (clebertsuconic) [inline artemis-commons/src/main/java/org/apache/activemq/artemis/utils/BoundedOrderedExecutorFactory.java]: I see where the word bounded came from.. there is a limit?

What about the Growable Queue?

File: artemis-commons/src/main/java/org/apache/activemq/artemis/utils/BoundedOrderedExecutorFactory.java (hunk 1 above)
Comment (franz1981) [inline artemis-commons/src/main/java/org/apache/activemq/artemis/utils/BoundedOrderedExecutorFactory.java]: I can change it :+1:
The main reason of that name is that the the max capacity (of submittable tasks) is bounded in some way...

File: artemis-commons/src/main/java/org/apache/activemq/artemis/utils/BoundedOrderedExecutorFactory.java (hunk 2 above)
Comment (franz1981) [inline artemis-commons/src/main/java/org/apache/activemq/artemis/utils/BoundedOrderedExecutorFactory.java]: Exactly!

**Comment by clebertsuconic** [comment]
What about simply change OrderedExecutorFactory /OrderedExecutor use the queue you want?
Maybe even pass in the queue as a parameter?


I'm sending a PR to an actorExecutor, if you could make your changes in top of that one?

**Comment by clebertsuconic** [comment]
so, if you could make a change to #1395 after merged? (if you agree with that PR), in such way that ProcessorBase would take the Queue type that you want.

**Pull request closed without merging**